hooks/useUpdatePoolInstances/index.ts

```diff
@@ -48,62 +51,78 @@ export const useUpdatePoolInstances = (): void => {
-                                setPoolsInitializationError(err);
-                            }
-                        })
-                        .finally(() => {
-                            setIsFetchingPools(false);
-                        });
-                };
-                fetchAndSetPools();
-            } else {
-                console.error('Skipped pools initialisation, network not supported');
-                setPoolsInitializationError(KnownPoolsInitialisationErrors.NetworkNotSupported);
-            }
-        } else {
-            console.error('Skipped pools initialisation, provider not ready');
+        if (isFetchingPools) {
+            console.debug('Skipped pools initialisation, already fetching pools');
+        } else if (!provider) {
+            console.debug('Skipped pools initialisation, provider not ready');
             resetPools();
             setPoolsInitializationError(KnownPoolsInitialisationErrors.ProviderNotReady);
+        } else if (!poolLists.length) {
+            console.debug('Skipped pools initialisation, poolList is empty');
+            setPoolsInitializationError(KnownPoolsInitialisationErrors.NoPools);
+        } else if (!network || !isSupportedNetwork(network)) {
+            console.debug(`Skipped pools initialisation, network: ${network} not supported`);
+            setPoolsInitializationError(KnownPoolsInitialisationErrors.NetworkNotSupported);
+        } else if (retryCount >= MAX_RETRY_COUNT) {
+            console.debug(`Skipped pools initialisation, retry count as exceeded ${MAX_RETRY_COUNT}`);
+            setPoolsInitializationError(KnownPoolsInitialisationErrors.ExceededMaxRetryCount);
+        } else {
+            // all is good
+            const fetchAndSetPools = async () => {
+                setIsFetchingPools(true);
+                console.debug(`Initialising pools ${network.slice()}: ${retryCount}`, poolLists);
+                resetPools();
+                hasSetPools.current = false;
+                setPoolsInitialized(false);
+                setPoolsInitializationError(undefined);
+
+                if (!poolLists.length) {
+                    return setPoolsInitializationError(KnownPoolsInitialisationErrors.NoPools);
+                }
+
+                const initialisedPools: Pool[] = [];
+
+                try {
+                    for (const pool of poolLists) {
+                        const poolSDKInstance = await attemptPromiseRecursively({
+                            promise: () =>
+                                Pool.Create({
+                                    ...pool,
+                                    address: pool.address,
+                                    provider,
+                                }),
+                            retryCheck: async () => {
+                                if (retryCount < MAX_RETRY_COUNT) {
+                                    setRetryCount(retryCount + 1);
+                                }
```

Comment:
Should we change this to a variable inside fetchAndSetPools instead of a state managed variable? Less fuckery between provider changes and poolLists changes

Reply:
yep good idea

Reply:
It would involve more work, but using a multicall to reduce the number of actual RPC requests would reduce throughput. Arbitrum has an official multicall contract `0x842eC2c7D803033Edf55E478F461FC547Bc54EB2`